aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -196,53 +196,99 @@ Tensor pinverse(const Tensor& self, double rcond) {
+              "torch.linalg.matrix_rank: rtol tensor of complex type is not supported.");
 
   // matrix_rank assigns a scalar value for each matrix in the batch so
-  // result's shape is equal to self.shape[0:self.ndim-2]
+  // result's shape is equal to input.shape[0:input.ndim-2]
   // for single matrix result_shape = {}
-  auto result_shape = IntArrayRef(self.sizes().cbegin(), self.sizes().cend()-2);
+  auto result_shape = IntArrayRef(input.sizes().cbegin(), input.sizes().cend() - 2);
   at::native::resize_output(result, result_shape);
 
   // NumPy doesn't take into account possible input with no elements and it errors on max not defined for this case
   // Let's output 0 for this case, since that kind of matrices have zero number of non-zero rows, hence rank is 0.
-  if (self.numel() == 0) {
+  if (input.numel() == 0) {
     result.fill_(0);
     return result;
   }
 
-  // We compute matrix rank as the number of singular or absolute eigen values above 'tol' threshold
-  Tensor S;
+  // We compute matrix rank as the number of singular or absolute eigen values
+  // that are above max(atol, rtol * max(S)) threshold
+  Tensor S, max_S;
   if (!hermitian) {
     Tensor U, V;
-    // TODO: replace self.svd with linalg_svd
-    std::tie(U, S, V) = self.svd(/*some=*/true, /*compute_uv=*/false);
+    // TODO: replace input.svd with linalg_svd
+    std::tie(U, S, V) = input.svd(/*some=*/true, /*compute_uv=*/false);
+    // singular values are sorted in descending order
+    max_S = at::narrow(S, /*dim=*/-1, /*start=*/0, /*length=*/1);
   } else {
-    S = at::linalg_eigvalsh(self);
+    S = at::linalg_eigvalsh(input);
     S = S.abs();
+    // eigenvalues are sorted in ascending order starting with negative values, we need a maximum value of abs(eigenvalues)
+    max_S = S.amax(/*dim=*/-1, /*keepdim=*/true);
   }
 
+  Tensor tol = at::max(atol.unsqueeze(-1), rtol * max_S);
+
+  result = at::sum_out(result, S > tol, /*dim=*/-1);
+  return result;
+}
+
+Tensor& linalg_matrix_rank_out(const Tensor& input, const Tensor& tol, bool hermitian, Tensor& result) {
+  // For NumPy compatibility tol is not scaled with max(singular_value) if the value for tol is provided
+  // It is assumed that the provided value is the absolute tolerance
+  Tensor rtol = at::zeros({}, tol.options());
+  result = linalg_matrix_rank_out_helper(input, tol, rtol, hermitian, result);
+  return result;
+}
+
+Tensor& linalg_matrix_rank_out(const Tensor& input, optional<double> tol, bool hermitian, Tensor& result) {
+  double tol_value;
+  Tensor atol, rtol;
   if (tol.has_value()) {
-    double tol_value = tol.value();
-    at::sum_out(result, S > tol_value, /*dim=*/-1);
+    tol_value = tol.value();
+    // For NumPy compatibility tol is not scaled with max(singular_value) if the value for tol is provided
+    // It is assumed that the provided value is the absolute tolerance
+    atol = at::full({}, tol_value, input.options().dtype(ScalarType::Double));
```

Comment:
What about casting to input's dtype?

Reply:
We get a number of type `double` from Python, I don't think we should cast it to input's dtype. There are separate. And in PyTorch it's possible to compare tensors with different types.
Same discussion for `torch.linalg.pinv`: https://github.com/pytorch/pytorch/pull/48399#discussion_r544443532